### PR TITLE
FI-2342: Moving the wallet connection logic to the Web SDK

### PR DESCRIPTION
**Pull request by samuel-kim-mesh**
https://www.notion.so/meshconnect/TDD-for-Updating-the-Web-SDK-for-Wagmi-45bc96b541e24d72b475cc100091e7fc?pvs=4

### CODE REVIEW

File: packages/link/src/utils/types.ts [hunk 1]
```diff
@@ -127,6 +135,18 @@ export interface LinkOptions {
    * Can be used to initialize the crypto transfers flow as an alternative to the target addresses.
    */
   transferDestinationTokens?: IntegrationAccessToken[]
+
+  /**
+   * (Optional) An array of Wagmi injected connector data. React hook
+   * It should send this list directly to our iframe.
+   */
+  injectedConnectors?: WagmiInjectedConnectorData[]
```
Comment (balint-front) [inline packages/link/src/utils/types.ts:143]: Please remove these parameters. The client's application does not need to provide us the available connectors.

File: packages/link/src/utils/wagmi.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  optimism
+} from 'wagmi/chains'
+import { walletConnect, coinbaseWallet, injected } from 'wagmi/connectors'
+
+export const config = createConfig({
+  chains: [mainnet, bsc, polygon, avalanche, arbitrum, optimism],
+  connectors: [
+    injected(),
+    walletConnect({
+      projectId: 'importWalletConnectId', //import correct walletConnectId
```
Comment (balint-front) [inline packages/link/src/utils/wagmi.ts:17]: Let's introduce a .env file, where we can configure this value.

File: packages/link/src/utils/wagmi.ts [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  mainnet,
+  polygon,
+  avalanche,
+  arbitrum,
+  optimism
+} from 'wagmi/chains'
+import { walletConnect, coinbaseWallet, injected } from 'wagmi/connectors'
+
+export const config = createConfig({
+  chains: [mainnet, bsc, polygon, avalanche, arbitrum, optimism],
```
Comment (balint-front) [inline packages/link/src/utils/wagmi.ts:13]: Let's ask Jason, what are the planned EVM chains, so we can include them as well, so we don't need a new SDK release when adding those

File: packages/link/src/Link.ts [hunk 4]
```diff
@@ -102,6 +102,16 @@ function eventsListener(
           iframeUrlObject?.origin || 'https://web.meshconnect.com'
         )
       }
+      if (currentOptions?.injectedCoreConnectors) {
+        //or use currentOptions?.injectedConnectors
+        iframeElement().contentWindow?.postMessage(
```
Comment (balint-front) [inline packages/link/src/Link.ts:107]: Let's introduce a helper function for sending messages to the Link UI, we will use it a lot.

File: packages/link/src/providers/WagmiCoreProvider.tsx [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import { FC, PropsWithChildren } from 'react'
+import { config } from '../utils/wagmiCore'
+import { WagmiProvider as WagmiProviderInternal } from 'wagmi'
+import { QueryClient, QueryClientProvider } from '@tanstack/react-query'
+
+export const WagmiCoreProvider: FC<PropsWithChildren> = ({ children }) => {
```
Comment (salavat-gainetdinov-mesh) [inline packages/link/src/providers/WagmiCoreProvider.tsx:6]: our link solutiuon should not depend on react, please don't use react or and tsx here

File: packages/link/src/utils/wagmiConnectorsUtils.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { useConnect } from 'wagmi'
+import { WagmiInjectedConnectorData } from './types'
+
+export const getWagmiInjectedData = (): WagmiInjectedConnectorData[] => {
```
Comment (salavat-gainetdinov-mesh) [inline packages/link/src/utils/wagmiConnectorsUtils.ts:4]: If you want to use react specific libs this methods should be a hook and name should be `useWagmiInjectedData`. But probably we don't want to depend on react here.

File: examples/react-example/src/ui/App.tsx [hunk 7]
```diff
@@ -135,21 +146,23 @@ export const App: React.FC = () => {
...
-        }}
-        onExit={err => {
-          setLinkToken(null)
-          setError(err || null)
-        }}
-        onTransferFinished={data => {
-          setTransferFinishedData(data)
-        }}
-      />
+      <WagmiCoreProvider>
```
Comment (salavat-gainetdinov-mesh) [inline examples/react-example/src/ui/App.tsx:149]: lets not force user to add more dependencies than needed.  link should be started by  `createLink` and `openLink` only

File: packages/link/src/Link.ts [hunk 8]
```diff
@@ -86,6 +96,95 @@ function eventsListener(
...
     }
+
+    case 'integrationInjectedWalletSelected': {
+      const payload = event.data.payload
+      console.log('Integration selected:', payload.integrationName)
+      console.log('Integration type', payload.integrationType)
+      const result: Connection = await connectToSpecificWallet(
+        payload.integrationName
+      )
+      console.log('result', result)
```
Comment (salavat-gainetdinov-mesh) [inline packages/link/src/Link.ts:107]: lets not add `console.log` in SDK

**Review by salavat-gainetdinov-mesh — CHANGES_REQUESTED** [review]

File: examples/react-example/src/ui/App.tsx [hunk 9]
```diff
@@ -22,9 +22,8 @@ export const App: React.FC = () => {
       }
     })
 
-    // this request should be performed from the backend side
```
Comment (salavat-gainetdinov-mesh) [inline examples/react-example/src/ui/App.tsx:25]: why did you removed comments here?

File: packages/link/src/Link.ts [hunk 10]
```diff
@@ -132,6 +151,204 @@ function eventsListener(
...
+    }
+  }
+}
+
+async function eventsListener(
+  event: MessageEvent<
+    LinkEventType | WalletBrowserEventType | { type: EventType }
+  >
+) {
+  if (isLinkEventTypeKey(event.data.type)) {
```
Comment (salavat-gainetdinov-mesh) [inline packages/link/src/Link.ts:341]: maybe its easier to do it like this:
```
if (isWalletBrowserEventTypeKey(event.data.type)) {
    await handleWalletBrowserEvent(
      event as MessageEvent<WalletBrowserEventType>
    )
  } else {
    await handleLinkEvent(event as MessageEvent<{ type: EventType }>)
  }
```

File: packages/link/package.json [hunk 11]
```diff
@@ -24,7 +24,11 @@
     "jest-environment-jsdom": "^29.7.0"
   },
   "dependencies": {
-    "@meshconnect/node-api": "^2.0.0"
+    "@meshconnect/node-api": "^2.0.0",
+    "@tanstack/react-query": "^5.17.9",
```
Comment (salavat-gainetdinov-mesh) [inline packages/link/package.json:28]: please remove this package

File: packages/link/src/index.ts [hunk 12]
```diff
@@ -1,3 +1,4 @@
 export * from './utils/types'
 export * from './utils/event-types'
+export * from './utils/wagmiCoreConnectorsUtils'
```
Comment (salavat-gainetdinov-mesh) [inline packages/link/src/index.ts:3]: Do users need to know about these utilities?

File: examples/react-example/src/ui/App.tsx
Comment (balint-front) [inline examples/react-example/src/ui/App.tsx:1]: Please revert all changes to this file before merging

File: packages/link/package.json [hunk 13]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@meshconnect/web-link-sdk",
-  "version": "2.0.5",
+  "version": "2.0.6",
```
Comment (balint-front) [inline packages/link/package.json:3]: Let's bump the minor version, so have it as 2.1.0

File: packages/link/package.json [hunk 14]
```diff
@@ -24,7 +24,10 @@
     "jest-environment-jsdom": "^29.7.0"
   },
   "dependencies": {
-    "@meshconnect/node-api": "^2.0.0"
+    "@tanstack/react-query": "^5.17.9",
+    "@wagmi/core": "^2.6.16",
+    "viem": "^2.x",
+    "wagmi": "^2.5.7"
```
Comment (balint-front) [inline packages/link/package.json:30]: Do we need this dependency? Seems like nothing is imported from this package

File: packages/link/src/Link.ts [hunk 15]
```diff
@@ -119,7 +132,14 @@ function eventsListener(
           payload: currentOptions.transferDestinationTokens
         })
       }
-
+      const injectedConnectors = await getWagmiCoreInjectedData()
+      console.log('injectedConnectors', injectedConnectors)
```
Comment (balint-front) [inline packages/link/src/Link.ts:136]: Pls remove logging

File: packages/link/src/Link.ts [hunk 16]
```diff
@@ -132,6 +152,202 @@ function eventsListener(
...
+          type: 'SDKinjectedConnected',
+          payload: {
+            accounts: result.accounts,
+            chainId: result.chainId,
+            signedTxHash: result.txSigned
+          }
+        })
+      } catch (error) {
+        sendMessageToIframe({
+          type: 'SDKinjectedDisconnected',
```
Comment (balint-front) [inline packages/link/src/Link.ts:173]: Can we call it something else? No disconnection happens here. Maybe `SDKinjectedConnectionError`

File: packages/link/src/Link.ts [hunk 17]
```diff
@@ -132,6 +152,202 @@ function eventsListener(
...
+        )
+        if (result instanceof Error) {
+          throw result
+        }
+        sendMessageToIframe({
+          type: 'SDKnativeTransferCompleted',
+          payload: result
+        })
+      } catch (error) {
+        console.error('Error sending native transaction:', error)
```
Comment (balint-front) [inline packages/link/src/Link.ts:230]: pls remove logging

File: packages/link/src/Link.ts [hunk 18]
```diff
@@ -132,6 +152,202 @@ function eventsListener(
...
+        sendMessageToIframe({
+          type: 'SDKnativeTransferCompleted',
```
Comment (balint-front) [inline packages/link/src/Link.ts:232]: Rename this event to `SDKnativeTransferError`

File: packages/link/src/Link.ts [hunk 19]
```diff
@@ -132,6 +152,202 @@ function eventsListener(
...
+          throw result
+        }
+        sendMessageToIframe({
+          type: 'SDKswitchChainSuccess',
+          payload: result
+        })
+      } catch (error) {
+        console.error('Error switching chain:', error)
+        sendMessageToIframe({
+          type: 'SDKswitchChainFailed',
```
Comment (balint-front) [inline packages/link/src/Link.ts:195]: Let's be consistent with event naming, and call this `SDKswitchChainError`

File: packages/link/src/Link.ts [hunk 20]
```diff
@@ -132,6 +152,202 @@ function eventsListener(
...
+          throw result
+        }
+        sendMessageToIframe({
+          type: 'SDKnonNativeTransferCompleted',
+          payload: result
+        })
+      } catch (error) {
+        console.error('Error sending non-native transaction:', error)
+        sendMessageToIframe({
+          type: 'SDKnonNativeTransferCompleted',
```
Comment (balint-front) [inline packages/link/src/Link.ts:259]: Same here with naming

File: packages/link/src/Link.ts [hunk 21]
```diff
@@ -132,6 +152,202 @@ function eventsListener(
...
+          throw result
+        }
+        sendMessageToIframe({
+          type: 'SDKnativeSmartDepositCompleted',
+          payload: {
+            txHash: result
+          }
+        })
+      } catch (error) {
+        console.error('Error sending native sc transaction:', error)
```
Comment (balint-front) [inline packages/link/src/Link.ts:287]: pls remove logging

File: packages/link/src/utils/wagmiCoreConnectorsUtils.ts [hunk 22]
```diff
@@ -0,0 +1,328 @@
...
+        type,
+        uid,
+        icon: icon || undefined
+      })
+    )
+    return mappedConnectors
+  })
+}
+
+type ConnectReturnTypeAndTxHash = {
```
Comment (balint-front) [inline packages/link/src/utils/wagmiCoreConnectorsUtils.ts:97]: Move this to types file

File: packages/link/src/utils/event-types.ts
Comment (balint-front) [inline packages/link/src/utils/event-types.ts:1]: Please move all of these new events to a new file called `wallet-connection-event-types.ts`

**Comment by balint-front** [comment]
Typecheck fails for this branch, please fix it

File: examples/react-example/src/ui/App.tsx [hunk 23]
```diff
@@ -135,7 +144,8 @@ export const App: React.FC = () => {
           Transfer
         </button>
       </div>
-
+      {/* <WagmiCoreProvider> */}
```
Comment (salavat-gainetdinov-mesh) [inline examples/react-example/src/ui/App.tsx]: please remove commented code here

File: packages/link/package.json (hunk 11 above)
Comment (samuel-kim-mesh) [reply packages/link/package.json]: `@tanstack/react-query` is already removed
and the latest `@meshconnect/node-api` is added

File: examples/react-example/src/ui/App.tsx (hunk 23 above)
Comment (samuel-kim-mesh) [reply examples/react-example/src/ui/App.tsx]: @salavat-gainetdinov-mesh please review the latest changes not the changes from yesterday.  Thank you.

**Review by salavat-gainetdinov-mesh — APPROVED** [review]

File: packages/link/src/Link.ts [hunk 24]
```diff
@@ -347,6 +304,21 @@ async function eventsListener(
   }
 }
 
+function handleErrorAndSendMessage(error: unknown, messageType: string) {
+  let errorMessage = 'An unexpected error occurred'
+  if (error instanceof Error) {
+    errorMessage = error.message
+  } else {
+    console.error('An unexpected error occurred', error)
```
Comment (balint-front) [inline packages/link/src/Link.ts:312]: Please don't print anything to the console. If you have this at other places, pls remove that

File: packages/link/package.json [hunk 25]
```diff
@@ -26,7 +27,8 @@
     "@meshconnect/node-api": "^2.0.5",
     "@wagmi/core": "^2.10.5",
     "viem": "^2.x",
-    "@wagmi/connectors": "^5.0.7"
+    "@wagmi/connectors": "^5.0.7",
+    "util": "^0.12.4"
```
Comment (balint-front) [inline packages/link/package.json:31]: I see util is only used for jest test, so it can be moved to devDependencies

**Review by balint-front — APPROVED** [review]

**Pull request merged**